File: factory_kernel/frontdoor_static/exploration.js
"use strict";
let explorationState = null;
let explorationPoll = null;
async function refreshExploration() {
  clearTimeout(explorationPoll);
  $("adaptive-card").hidden = !snapshot?.exploration_available;
  if (!snapshot?.exploration_available) return;
  try {
    explorationState = await api("/api/exploration");
    if (!token) return;
    renderExploration();
    if (explorationState.runs.some((run) => run.observation === "running")) {
      explorationPoll = setTimeout(() => {
        if (token) refreshExploration(); // Read-only observation; no automatic start or retry.
      }, 5000);
    }
  } catch (error) {
    explorationState = null;
    $("adaptive-open").hidden = true;
    $("adaptive-sessions").replaceChildren();
    $("adaptive-runs").replaceChildren();
    $("adaptive-state").textContent = "Exploration observation unavailable. Refresh before issuing another command.";
  }
}
function explorationCommand(session, request) {
  if (!explorationState) throw new Error("Refresh exploration before continuing.");
  return {idempotency_key: crypto.randomUUID(), expected_project_version: explorationState.project_version,
    session_id: session, request};
}
async function explorationAction(operation, session, request) {
  const command = explorationCommand(session, request);
  try {
    await api(`/api/exploration/${operation}`, command);
    message(operation === "start" ? "Bounded investigation recorded. Refresh and stop remain available." : "Exploration decision recorded.");
  } finally {
    // A temporarily unavailable observation must not be described as a failed command.
    try { await refresh(); } catch (error) {
      message("Command response observed, but current evidence is unavailable. Refresh before another command; no request was repeated.", true);
    }
  }
}
function explorationButton(parent, label, operation, session, request) {
  const button = text("button", label); button.type = "button";
  button.addEventListener("click", () => perform(() => explorationAction(operation, session, request)));
  parent.append(button);
}
function renderExploration() {
  const state = explorationState;
  const blocked = state.runs.some((run) => run.status === "running");
  const running = state.runs.some((run) => run.observation === "running");
  $("adaptive-state").textContent = state.ready ? "Planning evidence only · UNPROVEN. No work is published by exploring." : state.reason;
  if (state.budget) $("adaptive-state").textContent += ` Reserved ${state.budget.calls}/${state.budget.limits.calls} calls and $${state.budget.usd}/$${state.budget.limits.usd}.`;
  $("adaptive-open").hidden = !state.ready || blocked;
  const policy = $("adaptive-policy"); policy.replaceChildren();
  if (state.ready) {
    const frozen = state.default_policy;
    list(policy, "Comparison priorities, in order", frozen.priorities.map((id) => frozen.criteria.find((row) => row.id === id).question));
    policy.append(text("p", `Qualitative judgments; no measured performance claims. Shared limit for this approved scope: ${frozen.budget.calls} calls, $${frozen.budget.usd}, ${frozen.budget.probe_units} probe work units. Up to ${frozen.max_candidates} alternatives and ${frozen.max_rounds} investigation rounds.`));
    if (state.budget) policy.append(text("p", `Already reserved: ${state.budget.calls} calls and $${state.budget.usd}. ${state.budget.uncertain ? "Spend is uncertain; further paid work is frozen." : "Opening another question does not reset this budget."}`, "muted"));
  }
  const sessions = $("adaptive-sessions"); sessions.replaceChildren();
  for (const session of state.sessions) {
    const card = document.createElement("section");
    card.append(text("h3", session.question), text("p", `${session.status} · round ${session.round}`, "badge"));
    list(card, "Frozen comparison priorities", session.policy.priorities.map((id) => session.policy.criteria.find((row) => row.id === id).question));
    for (const outcome of (state.factory_outcomes || []).filter((row) => row.session_id === session.id)) {
      const observed = outcome.observation;
      card.append(text("h4", "Authenticated factory refusal"),
        text("p", `PR #${observed.receipt.pr}, run ${observed.receipt.run_id}, attempt ${observed.receipt.run_attempt}: ${observed.refusal.reason_code}.`),
        text("p", "The factory reported a refusal. Its cause is unresolved; this does not reject the strategy or qualify a replacement.", "muted"));
      const details = document.createElement("details"); details.append(text("summary", "Exact revision and evidence bindings"), text("pre", JSON.stringify(outcome, null, 2))); card.append(details);
    }
    if (session.handoffs.length) {
      const form = document.createElement("form");
      form.append(text("h4", "Import a factory outcome"), text("p", "Verify a current completed run against its retained evidence. Expired evidence, changed revisions and incomplete records are refused."));
      const fields = {};
      for (const [key, title] of [["run_id", "Workflow run ID"], ["attempt", "Run attempt"], ["pr", "Pull request number"]]) {
        const label = text("label", title); const input = document.createElement("input");
        input.type = "number"; input.min = "1"; input.step = "1"; input.required = true;
        if (key === "attempt") input.value = "1";
        label.append(input); form.append(label); fields[key] = input;
      }
      const label = text("label", "Programme item"); const item = document.createElement("select");
      for (const row of session.handoffs[session.handoffs.length - 1].input.proposal.items) {
        const option = text("option", row.id); option.value = row.id; item.append(option);
      }
      label.append(item); form.append(label);
      const submit = text("button", "Verify and import outcome"); submit.type = "submit"; form.append(submit);
      form.addEventListener("submit", (event) => {
        event.preventDefault();
        const request = Object.fromEntries(Object.entries(fields).map(([key, input]) => [key, Number(input.value)]));
        if (!Object.values(request).every(Number.isSafeInteger)) { message("Enter valid whole-number run and PR identities.", true); return; }
        perform(() => explorationAction("import-feedback", session.id, {...request, item_id: item.value}));
      });
      card.append(form);
    }
    for (const candidate of Object.values(session.candidates)) {
      const result = session.comparison.candidates[candidate.id];
      card.append(text("h4", candidate.mechanism), text("p", `${candidate.baseline ? "Baseline · " : ""}${result.status}`));
      for (const [criterion, value] of Object.entries(result.values)) {
        card.append(text("p", `${criterion}: ${value.assessment || `${value.low}–${value.high}`} (${value.kind}). ${value.basis}`));
      }
    }
    const recommendation = session.recommendations.at(-1);
    if (recommendation) {
      card.append(text("h4", "Recorded recommendation"), text("p", recommendation.rationale));
      list(card, "Remaining uncertainty", recommendation.remaining_uncertainty);
      card.append(text("p", `Next useful experiment: ${recommendation.next_useful_experiment}`));
    }
    if (session.handoffs.length) card.append(text("p", "A strategy handoff is recorded. Refresh evidence and use the separate publication review when it is current.", "muted"));
    const pending = Object.values(session.reservations).filter((row) => row.status === "pending");
    const handoffCurrent = session.handoff_current;
    if (!blocked && !pending.length && !handoffCurrent && !state.budget?.uncertain && ["exploring", "recommended"].includes(session.status)) {
      const form = document.createElement("form");
      const label = text("label", "Maximum reasoning steps (up to $1 reserved per call)");
      const steps = document.createElement("select");
      for (const count of [1, 2, 4, 8]) { const option = text("option", `${count} step${count === 1 ? "" : "s"} · up to $${count}`); option.value = count; steps.append(option); }
      label.append(steps); form.append(label, text("p", "The shared budget can stop this sooner. A result may require more evidence or an owner tradeoff; no automatic continuation.", "muted"));
      const submit = text("button", "Run bounded investigation"); submit.type = "submit"; form.append(submit);
      form.addEventListener("submit", (event) => { event.preventDefault(); perform(() => explorationAction("start", session.id, {max_steps: Number(steps.value), max_usd_per_call: 1})); });
      card.append(form);
    }
    if (!running) for (const reservation of pending) {
      card.append(text("p", "An earlier effect has no usable completion. Closing it keeps its full reservation charged and freezes further paid work when spend is unknown.", "muted"));
      explorationButton(card, "Close uncertain reservation without retry", "abandon", session.id,
        {reservation_id: reservation.id, reason: "Owner closed uncertain work without retry, refund or qualification."});
    }
    if (!blocked && !pending.length) explorationButton(card, "Reopen against current repository", "reopen", session.id,
      {reason: "Owner requested reconsideration against current repository context within the existing limits."});
    sessions.append(card);
  }
  const runs = $("adaptive-runs"); runs.replaceChildren();
  for (const run of state.runs.slice(-10).reverse()) {
    runs.append(text("p", `Investigation: ${run.observation}. ${run.reason || "Bounded work reserved; no completion confirmed yet."}`));
    if (state.ready && run.observation === "interrupted") explorationButton(runs,
      "Close interrupted job without restarting", "recover", run.session_id, {run_id: run.id});
  }
  setButtons();
}
$("adaptive-open").addEventListener("submit", (event) => {
  event.preventDefault();
  perform(async () => {
    const session = "question-" + crypto.randomUUID().replaceAll("-", "");
    await explorationAction("open", session, {question: $("adaptive-question").value,
      parent_session: null, policy: explorationState.default_policy});
    $("adaptive-question").value = ""; $("adaptive-policy-check").checked = false;
  });
});

File: factory_kernel/exploration.py
"""Adaptive, owner-scoped exploration before programme commitment. Never qualification."""
from __future__ import annotations

from copy import deepcopy

from .canonical import sha256_value
from .frontdoor_intent import IntentRefused, _shape, _text, _texts
from .frontdoor_programme import prepare_programme
from .programme import compile_programme
from .exploration_policy import comparison, validate_addition, validate_policy, validate_predictions
from .exploration_probe import METRICS, run_probe, validate_probe
from .exploration_records import ExplorationRecords


class Exploration:
    def __init__(self, store, context, *, check_stop, app_login):
        self.records = ExplorationRecords(store)
        self.context, self.check_stop, self.app_login = context, check_stop, app_login

    def _context(self):
        value = deepcopy(self.context())
        _shape(value, {"commit", "files", "policies", "coverage", "proof_status", "identity"})
        if value["identity"] != sha256_value({key: val for key, val in value.items() if key != "identity"}):
            raise IntentRefused("repository analysis identity mismatch")
        return value

    def _session(self, state, approval, key, *, allow_stale=False):
        session = state["sessions"].get(key)
        if session is None or session["binding"]["spec_sha256"] != approval["spec_sha256"]:
            raise IntentRefused("session does not bind current approved scope")
        if session["binding"]["approval_version"] != approval["project_version"]:
            raise IntentRefused("session approval has been superseded")
        if not allow_stale and session["context"] != self._context():
            raise IntentRefused("repository changed; reconsider before further work")
        return session

    def _append(self, project, principal, command, action, transition):
        self.check_stop()
        command = deepcopy(command)
        command["request"] = {"action": action, "data": command["request"]}
        def replay_guard(state, approval, recorded):
            session = self._session(state, approval, command["session_id"],
                          allow_stale=action in {"observe-claim", "reopen", "abandon-pending"})
            if action == "handoff" and (session["status"] != "recommended" or
                    recorded["data"]["recommendation_sha256"] != sha256_value(session["recommendations"][-1])):
                raise IntentRefused("recorded handoff no longer has a current recommendation")
        return self.records.append(project, principal, command, transition, replay_guard)

    def open(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            _shape(request, {"question", "parent_session", "policy"})
            _text(request["question"], 4000)
            validate_policy(request["policy"])
            if command["session_id"] in state["sessions"] or len(state["sessions"]) >= 40:
                raise IntentRefused("session already exists or project search capacity reached")
            if request["parent_session"] is not None:
                self._session(state, approval, request["parent_session"])
            previous = state["budgets"].get(approval["spec_sha256"])
            if previous and previous["limits"] != request["policy"]["budget"]:
                raise IntentRefused("all questions share the frozen approved-scope budget")
            return "opened", {**request, "context": self._context(), "binding": {
                "spec_sha256": approval["spec_sha256"], "approval_version": approval["project_version"]}}

        return self._append(project, principal, command, "open", transition)[0]

    def add_candidates(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"])
            if session["status"] != "exploring":
                raise IntentRefused("reopen the question before adding alternatives")
            validate_addition(request, state, session, approval["spec"])
            return "candidates-added", request

        return self._append(project, principal, command, "add-candidates", transition)[0]

    def inspect(self, project, session_id, *, principal):
        state, approval = self.records.read(project, principal)
        session = self._session(state, approval, session_id)
        result = comparison(state, session)
        budget = state["budgets"][approval["spec_sha256"]]
        return {"project_version": state["project_version"], "session": deepcopy(session),
                "comparison": result, "budget": budget,
                "next_action": "recommend-or-investigate" if result["sufficient_support"] else
                               "investigate-decision-changing-uncertainty" if result["preferred"] else
                               "generate-different-strategy"}

    def assess(self, project, command, *, principal):
        """Retain revised predictions/judgments; they can never overwrite measurements."""
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"])
            _shape(request, {"candidates", "basis"})
            _text(request["basis"], 4000)
            if session["status"] != "exploring" or not session["candidates"]:
                raise IntentRefused("assess an existing open candidate pool")
            _shape(request["candidates"], session["candidates"])
            criteria = {row["id"]: row for row in session["policy"]["criteria"]}
            for predictions in request["candidates"].values():
                validate_predictions(predictions, criteria)
            return "assessed", {**request, "round": session["round"], "kind": "reasoned-assessment",
                "context_identity": session["context"]["identity"], "qualification_status": "UNPROVEN"}

        return self._append(project, principal, command, "assess", transition)[0]

    def experiment(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def reserve(state, approval):
            session = self._session(state, approval, command["session_id"])
            if session["status"] != "exploring":
                raise IntentRefused("experiments require an open question")
            _shape(request, {"probe", "targets", "question", "would_change_decision_if", "claim_ids"})
            for field in ("question", "would_change_decision_if"):
                _text(request[field], 4000)
            claims = _texts(request["claim_ids"])
            if not claims or not set(claims) <= state["claims"].keys():
                raise IntentRefused("experiment must investigate recorded claims")
            units = validate_probe(request["probe"])
            criteria = {row["id"]: row for row in session["policy"]["criteria"]}
            if not isinstance(request["targets"], list) or not 1 <= len(request["targets"]) <= 30:
                raise IntentRefused("experiment needs bounded pre-registered measurement targets")
            seen = set()
            for target in request["targets"]:
                _shape(target, {"candidate_id", "criterion_id", "metric", "falsifies_claim"})
                candidate = session["candidates"].get(target["candidate_id"])
                criterion = criteria.get(target["criterion_id"])
                identity = (target["candidate_id"], target["criterion_id"])
                if (candidate is None or criterion is None or criterion["kind"] != "measurement" or target["metric"] not in METRICS
                        or criterion["unit"] != target["metric"] or identity in seen
                        or candidate["probe_strategy"] not in request["probe"]["strategies"]
                        or not set(candidate["claim_ids"]).intersection(claims)
                        or (target["falsifies_claim"] is not None and (criterion["ceiling"] is None
                            or target["falsifies_claim"] not in candidate["claim_ids"]
                            or target["falsifies_claim"] not in claims))):
                    raise IntentRefused("measurement target or unit is not bound to this experiment")
                seen.add(identity)
            budget = state["budgets"][approval["spec_sha256"]]
            if budget["probe_units"] + units > budget["limits"]["probe_units"]:
                raise IntentRefused("shared cumulative experiment budget exhausted")
            identity = sha256_value({"session": session["id"], "round": session["round"], "request": request})
            if identity in session["reservations"]:
                raise IntentRefused("this experiment was already reserved; do not repeat uncertain work")
            return "reserved", {"id": identity, "calls": 0, "usd": 0, "probe_units": units,
                "request": request, "round": session["round"], "context_identity": session["context"]["identity"]}

        state, event, created = self._append(project, principal, command, "experiment", reserve)
        if not created:
            return state
        reservation = event["data"]
        # No model-supplied runner or callback is accepted. Only the fixed data-only runner.
        try:
            receipt = run_probe(request["probe"], check_stop=self.check_stop)
            status, failure = "complete", None
        except Exception as exc:
            receipt, status, failure = None, "failed", type(exc).__name__
        completion = {"idempotency_key": "probe-result-" + reservation["id"],
                      "expected_project_version": state["project_version"],
                      "session_id": command["session_id"], "request": {"reservation": reservation["id"]}}

        def finish(current, approval):
            session = self._session(current, approval, command["session_id"])
            pending = session["reservations"][reservation["id"]]
            if pending["status"] != "pending" or pending["round"] != session["round"]:
                raise IntentRefused("experiment result is stale or already recorded")
            measurements = []
            claim_observations = []
            if receipt is not None:
                ceilings = {row["id"]: row["ceiling"] for row in session["policy"]["criteria"]}
                for target in request["targets"]:
                    candidate = session["candidates"][target["candidate_id"]]
                    value = receipt["results"][candidate["probe_strategy"]][target["metric"]]
                    measurements.append({**target, "value": value})
                    if target["falsifies_claim"] is not None:
                        claim_observations.append({"claim_id": target["falsifies_claim"],
                            "outcome": "contradicted" if value > ceilings[target["criterion_id"]] else "supported-in-probe",
                            "receipt_sha256": sha256_value(receipt), "context_identity": session["context"]["identity"],
                            "qualification_status": "UNPROVEN", "scope": "finite-declared-workload"})
            return "observed", {"reservation_id": reservation["id"], "round": session["round"],
                "context_identity": session["context"]["identity"], "status": status, "failure": failure,
                "receipt": receipt, "receipt_sha256": sha256_value(receipt), "measurements": measurements,
                "claim_observations": claim_observations}

        # A stopped/stale/concurrent completion remains pending and charged. Never rerun it.
        return self._append(project, principal, completion, "experiment-result", finish)[0]

    def recommend(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"])
            _shape(request, {"stop_reason", "rationale", "remaining_uncertainty", "next_useful_experiment"})
            _text(request["rationale"], 4000)
            _texts(request["remaining_uncertainty"])
            _text(request["next_useful_experiment"], 4000)
            if session["status"] != "exploring" or len({c["family"] for c in session["candidates"].values()}) < 2:
                raise IntentRefused("recommendation requires an open question and different strategy families")
            if any(row["status"] == "pending" for row in session["reservations"].values()):
                raise IntentRefused("uncertain reserved work cannot be hidden by a recommendation")
            result = comparison(state, session)
            if result["preferred"] is None:
                raise IntentRefused("no viable strategy; reconsider the candidate pool")
            if request["stop_reason"] == "sufficient-support":
                if not result["sufficient_support"]:
                    raise IntentRefused("evidence intervals still leave a decision-changing tradeoff")
            elif request["stop_reason"] == "bounded-decision":
                if not request["remaining_uncertainty"]:
                    raise IntentRefused("bounded stopping must disclose remaining uncertainty")
            else:
                raise IntentRefused("unknown search stopping rule")
            candidate = session["candidates"][result["preferred"]]
            return "recommended", {**request, "candidate_id": candidate["id"], "claim_ids": candidate["claim_ids"],
                "round": session["round"], "comparison": result, "context_identity": session["context"]["identity"],
                "candidate_sha256": sha256_value(candidate), "proof_reuse_allowed": False}

        return self._append(project, principal, command, "recommend", transition)[0]

    def abandon_pending(self, project, command, *, principal):
        """Close uncertain work without retrying it or releasing any reserved budget."""
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"], allow_stale=True)
            _shape(request, {"reservation_id", "reason"})
            _text(request["reason"], 4000)
            reservation = session["reservations"].get(request["reservation_id"])
            if reservation is None or reservation["status"] != "pending":
                raise IntentRefused("only pending work can be abandoned")
            return "observed", {"reservation_id": request["reservation_id"], "status": "abandoned",
                "round": reservation["round"], "context_identity": reservation["context_identity"],
                "reason": request["reason"], "reported_usd": None, "measurements": [],
                "receipt_sha256": sha256_value(request), "qualification_status": "UNPROVEN"}

        return self._append(project, principal, command, "abandon-pending", transition)[0]

    def observe_claim(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            self._session(state, approval, command["session_id"], allow_stale=True)
            _shape(request, {"claim_id", "status", "observation", "source"})
            if request["claim_id"] not in state["claims"] or request["status"] not in {"challenged", "invalidated"}:
                raise IntentRefused("unknown claim or invalid observation status")
            if state["claims"][request["claim_id"]]["status"] == "invalidated":
                raise IntentRefused("invalidated claims are immutable; record a revised claim")
            for field in ("observation", "source"):
                _text(request[field], 4000)
            return "claim-observed", {**request, "evidence_class": "owner-reported", "qualification_status": "UNPROVEN"}

        return self._append(project, principal, command, "observe-claim", transition)[0]

    def reopen(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"], allow_stale=True)
            _shape(request, {"reason"})
            _text(request["reason"], 4000)
            if session["round"] >= session["policy"]["max_rounds"]:
                raise IntentRefused("reconsideration round budget exhausted")
            if any(row["status"] == "pending" for row in session["reservations"].values()):
                raise IntentRefused("pending effects require reconciliation before reconsideration")
            return "reopened", {**request, "context": self._context()}

        return self._append(project, principal, command, "reopen", transition)[0]

    def handoff(self, project, command, *, principal):
        request = deepcopy(command["request"])

        def transition(state, approval):
            session = self._session(state, approval, command["session_id"])
            _shape(request, {"proposal"})
            if session["status"] != "recommended":
                raise IntentRefused("a current recommendation is required")
            recommendation = session["recommendations"][-1]
            value = {"version": "1.0", "spec": deepcopy(approval["spec"]),
                     "proposal": request["proposal"], "app_login": self.app_login}
            programme = compile_programme(value, repository=self.records.store.repository)
            return "handoff", {"schema": "dark-factory/exploration-handoff", "schema_version": "1.0",
                "input": value, "input_sha256": sha256_value(value), "programme_sha256": programme.sha256,
                "recommendation_sha256": sha256_value(recommendation), "candidate_id": recommendation["candidate_id"],
                "strategy": deepcopy(session["candidates"][recommendation["candidate_id"]]),
                "claim_ids": recommendation["claim_ids"], "context_identity": session["context"]["identity"],
                "binding": session["binding"], "qualification_status": "UNPROVEN", "proof_reuse_allowed": False,
                "activation": "requires-normal-publication-admission-and-fresh-qualification"}

        return self._append(project, principal, command, "handoff", transition)[0]

    def prepare_handoff(self, project, session_id, *, expected_project_version, principal):
        """Revalidate stored recommendation and use the existing review/admission input seam.

        This exports a proposal, never a publication capability. The publisher must still
        obtain its normal exact-input owner consent and recheck currency before effects.
        """
        self.check_stop()
        state, approval = self.records.read(project, principal)
        if type(expected_project_version) is not int or expected_project_version != state["project_version"]:
            raise IntentRefused("stale project version before handoff preparation")
        session = self._session(state, approval, session_id)
        if session["status"] != "recommended" or not session["handoffs"]:
            raise IntentRefused("no current stored exploration handoff")
        handoff = session["handoffs"][-1]
        if handoff["recommendation_sha256"] != sha256_value(session["recommendations"][-1]):
            raise IntentRefused("handoff recommendation was superseded")
        review = prepare_programme(self.records.store, project, {
            "expected_project_version": expected_project_version, "approval_version": approval["project_version"],
            "spec_sha256": approval["spec_sha256"], "proposal": deepcopy(handoff["input"]["proposal"])},
            principal=principal, app_login=self.app_login)
        if review["input_sha256"] != handoff["input_sha256"] or review["programme_sha256"] != handoff["programme_sha256"]:
            raise IntentRefused("stored handoff does not match the current programme compiler")
        self.check_stop()
        if self._context() != session["context"]:
            raise IntentRefused("repository changed during handoff preparation")
        return {**review, "exploration": {"session_id": session_id, "handoff_sha256": sha256_value(handoff),
            "recommendation_sha256": handoff["recommendation_sha256"], "strategy": deepcopy(handoff["strategy"]),
            "claim_ids": handoff["claim_ids"], "qualification_status": "UNPROVEN", "proof_reuse_allowed": False,
            "strategy_enforcement": "advisory-sidecar-not-consumed-by-current-factory-workers"}}
